dbms/src/Interpreters/JoinV2/HashJoin.cpp

```diff
@@ -0,0 +1,805 @@
+        "Logical error, the output of join contains duplicated columns");
+
+    output_block_after_finalize = Block(output_columns_after_finalize);
+    Names updated_require;
+    if (match_helper_name.empty())
+        updated_require = parent_require;
+    else
+    {
+        required_names_set.erase(match_helper_name);
+        for (const auto & name : required_names_set)
+            updated_require.push_back(name);
+    }
+    if (!non_equal_conditions.null_aware_eq_cond_name.empty())
+    {
+        updated_require.push_back(non_equal_conditions.null_aware_eq_cond_name);
+    }
+    if (!non_equal_conditions.other_eq_cond_from_in_name.empty())
+        updated_require.push_back(non_equal_conditions.other_eq_cond_from_in_name);
+    if (!non_equal_conditions.other_cond_name.empty())
+        updated_require.push_back(non_equal_conditions.other_cond_name);
+    /// join will reuse the input columns so need to let finalize keep the input columns
+    if (non_equal_conditions.null_aware_eq_cond_expr != nullptr)
+    {
+        non_equal_conditions.null_aware_eq_cond_expr->finalize(updated_require, true);
+        updated_require = non_equal_conditions.null_aware_eq_cond_expr->getRequiredColumns();
+    }
+    if (non_equal_conditions.other_cond_expr != nullptr)
+    {
+        non_equal_conditions.other_cond_expr->finalize(updated_require, true);
+        updated_require = non_equal_conditions.other_cond_expr->getRequiredColumns();
+    }
+
+    if (non_equal_conditions.other_cond_expr != nullptr || non_equal_conditions.null_aware_eq_cond_expr != nullptr)
+    {
+        output_columns_names_set_for_other_condition_after_finalize = output_column_names_set_after_finalize;
+        for (const auto & name : updated_require)
+            output_columns_names_set_for_other_condition_after_finalize.insert(name);
+        if (!match_helper_name.empty())
+            output_columns_names_set_for_other_condition_after_finalize.insert(match_helper_name);
+    }
+
+    /// remove duplicated column
+    required_names_set.clear();
+    for (const auto & name : updated_require)
+        required_names_set.insert(name);
+    /// add some internal used columns
+    if (!non_equal_conditions.left_filter_column.empty())
+        required_names_set.insert(non_equal_conditions.left_filter_column);
+    if (!non_equal_conditions.right_filter_column.empty())
+        required_names_set.insert(non_equal_conditions.right_filter_column);
+    /// add join key to required_columns
+    for (const auto & name : key_names_right)
+        required_names_set.insert(name);
+    for (const auto & name : key_names_left)
+        required_names_set.insert(name);
+
+    for (const auto & name : required_names_set)
+        required_columns.push_back(name);
+    finalized = true;
+}
+
+Block HashJoin::handleOtherConditions(size_t stream_index)
+{
+    auto & wd = probe_workers_data[stream_index];
+    non_equal_conditions.other_cond_expr->execute(wd.result_block);
```

Comment:
question: HashJoinV2 still needs to materialize **all columns** and compute other conditions, instead of only materialize columns that needed by other condition?

Reply:
Yes, I will do that later.